### PR TITLE
Allow setting a custom pixel ratio (#769)

### PR DESCRIPTION
**Pull request by vanilla-lake**
## Launch Checklist

<changelog>Allow setting a custom pixel ratio by adding a `MapOptions#pixelRatio` property and a `Map#setPixelRatio` method (#769)</changelog>

These changes allow setting a custom pixel ratio, overriding `devicePixelRatio`. This can be useful when exporting an image of the map at a different resolution than the in-app map while still preserving the same extent and zoom level. Two new methods are added on `Map`: `setPixelRatio` and `getPixelRatio`. In addition, a `pixelRatio` property is added to `MapOptions`.

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [ ] Manually test the debug page.
 - [x] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog></changelog>`.

### CODE REVIEW

**Review by HarelM — APPROVED** [review]
This looks good code wise, instead of using a global variable the code now uses a member. I think this is a good improvement.

**Comment by HarelM** [comment]
@wipfli @astridx @xabbu42 I think someone else other than me should review it.
Overall this looks very good.
Thanks a lot @vanilla-lake !!

**Comment by vanilla-lake** [comment]
My pleasure. Thanks for the review!

**Comment by wipfli** [comment]
Thanks @vanilla-lake, your code changes look good to me too. Please add a render test for this new functionality: https://github.com/maplibre/maplibre-gl-js/blob/main/test/integration/README.md#writing-new-tests

File: src/ui/map.ts [hunk 1]
```diff
@@ -298,6 +300,8 @@ class Map extends Camera {
     _locale: any;
     _removed: boolean;
     _clickTolerance: number;
+    _pixelRatio: number;
+    _styleSpec: StyleSpecification | string;
```
Comment (HarelM) [inline src/ui/map.ts:304]: Style can be set and then alerted, keeping the original version is fragile...
I would not add the update style and let the developer take care of this.

**Comment by vanilla-lake** [comment]
@wipfli I just changed render and query tests so their `pixelRatio` option uses the new `MapOptions#pixelRatio` option. I didn't add any tests, as there are already some testing different pixel ratios and the new ones would most likely have been redundant. Please let me know if this looks okay to you.

**Comment by astridx** [comment]
@vanilla-lake You wrote in #769:

> ... However, since not all browsers display a warning, maybe we could display one ourselves when the canvas' width or height exceed the value returned by `gl.getParameter(gl.MAX_RENDERBUFFER_SIZE)`?

This must already exist or have existed. I found this line in the changelog:

https://github.com/maplibre/maplibre-gl-js/blob/809b1c93c797f5e3ecae1191e8e7fa7d90dd6995/CHANGELOG.md?plain=1#L1478

**Comment by vanilla-lake** [comment]
@astridx They were removed as part of https://github.com/mapbox/mapbox-gl-js/pull/4207. I think the problem was that they divided the value returned by `gl.getParameter(gl.MAX_RENDERBUFFER_SIZE)` by 2 (perhaps because of some confusion when `devicePixelRatio == 2`). In my tests, exceeding the size returned by `gl.getParameter(gl.MAX_RENDERBUFFER_SIZE)` caused the issue to happen, but it worked fine for sizes less than that.

**Comment by astridx** [comment]
> @astridx They were removed as part of [mapbox/mapbox-gl-js#4207](https://github.com/mapbox/mapbox-gl-js/pull/4207). I think the problem was that they divided the value returned by `gl.getParameter(gl.MAX_RENDERBUFFER_SIZE)` by 2 (perhaps because of some confusion when `devicePixelRatio == 2`). In my tests, exceeding the size returned by `gl.getParameter(gl.MAX_RENDERBUFFER_SIZE)` caused the issue to happen, but it worked fine for sizes less than that.


What do you mean by that exactly? Dividing by 2 is the most correct value, or have I not understood that correctly? 

If I understand correctly, there have been problems in the past on different devices. But since that was years ago, it may have changed. I am on Linux and here I get a warning in the console in Firefox and in Chrome with the following code and the map section is too small. That is OK.


        const gl = this.painter.context.gl;
        const maxSize = gl.getParameter(gl.MAX_RENDERBUFFER_SIZE) / 2;
        
        if (this._canvas.width > maxSize || this._canvas.height > maxSize) {
            performance.warnOnce(
                `Mapibre Map is larger than maximum size supported by this system ` +
                `(${maxSize}px).`
            );
        }


I would merge this PR as it is and include the warning (something like the previous code) in a following step and test this warning on different devices. For this, whoever merges the PR would have to open an issue in this repo.  We would also need to open an issue in the documentation repo so that we don't forget to document.

What do others think?

**Comment by vanilla-lake** [comment]
@astridx That is strange. When I tested on macOS, the limit appeared to be the value returned by `gl.getParameter(gl.MAX_RENDERBUFFER_SIZE)`, not `gl.getParameter(gl.MAX_RENDERBUFFER_SIZE) / 2`. I also don't understand why we would need to divide by 2. https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glGet.xml states:
> GL_MAX_RENDERBUFFER_SIZE
params returns one value. The value indicates the largest renderbuffer width and height that the GL can handle. The value must be at least 1. See glRenderbufferStorage.

I agree that it may be best to keep that for a later PR.

**Comment by astridx** [comment]
As I mentioned in https://github.com/maplibre/maplibre-gl-js/issues/769#issuecomment-1015338010, we should add this feature. The problems I had there have nothing to do with this PR directly. Since others have also commented positively here and no one has anything against it, I am now merging the PR. I hope this is fine.
Thank you very much @vanilla-lake

**Comment by vanilla-lake** [comment]
@astridx No problem! Thank you for reviewing this PR!

**Comment by astridx** [comment]
> @astridx That is strange. When I tested on macOS, the limit appeared to be the value returned by `gl.getParameter(gl.MAX_RENDERBUFFER_SIZE)`, not `gl.getParameter(gl.MAX_RENDERBUFFER_SIZE) / 2`. I also don't understand why we would need to divide by 2. https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glGet.xml states:


I think you are right. Until now, it was not clear to me what `width **and** height `in `'GL_MAX_RENDERBUFFER_SIZE = params returns one value. The value indicates the largest renderbuffer width and height that the GL can handle. The value must be at least 1. '` (https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glGet.xml) means. 
[Here](https://www.khronos.org/opengl/wiki/Texture#Therorie)it is clearer for the parameter `GL_MAX_TEXTURE_SIZE `from my point of view:

> ... For 1D and 2D textures (and any texture types that use similar dimensionality, like cubemaps) the max size of either dimension is GL_MAX_TEXTURE_SIZE. ...

The same certainly applies analogously to `MAX_RENDERBUFFER_SIZE`. However, I did not find this directly in the wiki and since I already had the problems with smaller dimensions, I did not investigate this further. 
Probably the specifics that apply to WebGL on any [device/browser](https://stackoverflow.com/questions/59510450/webgl-parameters-not-consistent-between-browsers) are the reason.

**Pull request merged**